From 01b5f36a5ca62a77dddf1b22c7e1eac910d96870 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Thu, 31 Oct 2013 14:19:58 +0000
Subject: [PATCH 027/351] extcon: arizona: Fix headphone detection accuracy on
 wm5110

wm5110 requires slightly different configuration of the headphone
detection to other Arizona devices. Otherwise accuracy will be limited
especially at high impedances.

Change-Id: I706928ca4201f698659f253e254c31bb5d728d27
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 drivers/extcon/extcon-arizona.c |   22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/drivers/extcon/extcon-arizona.c b/drivers/extcon/extcon-arizona.c
index 41cc25a..83af0f7 100644
--- a/drivers/extcon/extcon-arizona.c
+++ b/drivers/extcon/extcon-arizona.c
@@ -145,8 +145,24 @@ static void arizona_extcon_do_magic(struct arizona_extcon_info *info,
 				    unsigned int magic)
 {
 	struct arizona *arizona = info->arizona;
+	unsigned int mask = 0, val = 0;
 	int ret;
 
+	switch (arizona->type) {
+	case WM5110:
+		mask = 0x0007;
+		if (magic)
+			val = 0x0001;
+		else
+			val = 0x0006;
+		break;
+	default:
+		mask = 0x4000;
+		if (magic)
+			val = 0x4000;
+		break;
+	};
+
 	mutex_lock(&arizona->dapm->card->dapm_mutex);
 
 	arizona->hpdet_magic = magic;
@@ -163,14 +179,12 @@ static void arizona_extcon_do_magic(struct arizona_extcon_info *info,
 				 ret);
 	}
 
-	ret = regmap_update_bits(arizona->regmap, 0x225, 0x4000,
-				 magic);
+	ret = regmap_update_bits(arizona->regmap, 0x225, mask, val);
 	if (ret != 0)
 		dev_warn(arizona->dev, "Failed to do magic: %d\n",
 				 ret);
 
-	ret = regmap_update_bits(arizona->regmap, 0x226, 0x4000,
-				 magic);
+	ret = regmap_update_bits(arizona->regmap, 0x226, mask, val);
 	if (ret != 0)
 		dev_warn(arizona->dev, "Failed to do magic: %d\n",
 			 ret);
-- 
1.7.9.5

